### PR TITLE
Create SteamKit logger to log packets in nethook format

### PR DESCRIPTION
**Pull request by xPaw**
This is initial implementation to gather feedback. What do you think?

### CODE REVIEW

**Comment by VoiDeD** [comment]
Tossing some thoughts along with some from the IRC discussion:

1. I don't want to lock this down to one specific file format (filename, directory, etc), but rather allow consumers implement something like an `INetworkListener` interface, and let them implement the logic for where dumped traffic should go.

2. I don't think we should expose this from the constructor, it might give consumers the idea that it's a first class feature of `SteamClient`, rather than what it really is: a debugging convenience. So on that topic, something like adding a `SetNetworkListener` on `CMClient` would likely be more ideal.

3. Also don't want to make it convenient for consumers to do something like:
```c#
public void OnIncomingMessage( EMsg msgType, byte[] data )
{
    if ( msgType == EMsg.ClientLogOnResponse )
    {
        // business logic...
    }
}
```
So what if we made this a DEBUG build only feature?

**Comment by xPaw** [comment]
I agree with first two, however I think the nethook implementation should be kept as well for ease of use with nethook analyzer.

File: SteamKit2/SteamKit2/Steam/CMClient.cs [hunk 1]
```diff
@@ -189,6 +192,10 @@ public void Send( IClientMsg msg )
 
             DebugLog.WriteLine( "CMClient", "Sent -> EMsg: {0} (Proto: {1})", msg.MsgType, msg.IsProto );
 
+            if ( netHookLogger != null )
+            {
+                netHookLogger.LogOutgoingNetMessage( msg.MsgType, msg.Serialize() );
```
Comment (Netshroud) [inline SteamKit2/SteamKit2/Steam/CMClient.cs]: Use C# 6 null-conditional operator. :)

File: SteamKit2/SteamKit2/Util/NetHookLogger.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        private uint MessageNumber;
+        private string LogDirectory;
+
+        /// <summary>
+        /// Will create a folder in path "%assembly%/nethook/%currenttime%/"
+        /// </summary>
+        public NetHookLogger()
+        {
+            LogDirectory = Path.Combine(
+                Path.GetDirectoryName( System.Reflection.Assembly.GetExecutingAssembly().Location ),
```
Comment (Netshroud) [inline SteamKit2/SteamKit2/Util/NetHookLogger.cs]: Use `CodeBase`, not `Location`.

**Comment by Netshroud** [comment]
I agree with @VoiDeD, but I do think a NetHook-compatible implementation would be useful - both for myself, as well as a reference implementation for that interface.

I agree with point 3, but think this would also be useful in Release builds (e.g. to monitor and report certain unhandled message types such as VAC). I'm not sure how else to prevent if-else handling by consumers, though.

File: Samples/3.DebugLog/Program.cs [hunk 3]
```diff
@@ -75,7 +75,9 @@ static void Main( string[] args )
             pass = args[ 1 ];
 
             // create our steamclient instance
-            steamClient = new SteamClient();
+            steamClient = new SteamClient( System.Net.Sockets.ProtocolType.Tcp );
```
Comment (VoiDeD) [inline Samples/3.DebugLog/Program.cs]: Nit: Don't need to modify the constructor call here.

File: SteamKit2/SteamKit2/Util/DebugNetworkListener.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        /// <param name="msgType">Network message type of this packet message.</param>
+        /// <param name="data">Raw packet data that will be sent.</param>
+        void OnOutgoingNetMessage( EMsg msgType, byte[] data );
+    }
+
+    /// <summary>
+    /// Dump any network messages sent to and received from the Steam server that the client is connected to.
+    /// These messages are dumped to file, and can be analyzed further with NetHookAnalyzer, a hex editor, or your own purpose-built tools.
+    /// </summary>
+    public class NetHookLogger : IDebugNetworkListener
```
Comment (VoiDeD) [inline SteamKit2/SteamKit2/Util/DebugNetworkListener.cs]: What about calling it `NetHookNetworkListener`?

File: Samples/3.DebugLog/Program.cs (hunk 3 above)
Comment (xPaw) [inline Samples/3.DebugLog/Program.cs]: Unsure why it got changed, will fix.

**Comment by VoiDeD** [comment]
This is probably fine to introduce unless @Netshroud has any other reservations.

The only last consideration is any security exceptions from `NetHookNetworkListener` attempting to write to a directory without permissions.

File: Samples/3.DebugLog/Program.cs [hunk 5]
```diff
@@ -76,6 +76,8 @@ static void Main( string[] args )
 
             // create our steamclient instance
             steamClient = new SteamClient();
+            steamClient.DebugNetworkListener = new NetHookNetworkListener();
```
Comment (Netshroud) [inline Samples/3.DebugLog/Program.cs]: Is this sample really the place for this? Users might have their passwords stored locally in dumps and not realise it.

File: SteamKit2/SteamKit2/Util/DebugNetworkListener.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    /// This interface is used for logging network messages sent to and received from the Steam server that the client is connected to.
+    /// </summary>
+    public interface IDebugNetworkListener
+    {
+        /// <summary>
+        /// Called when a packet is received from the Steam server.
+        /// </summary>
+        /// <param name="msgType">Network message type of this packet message.</param>
+        /// <param name="data">Raw packet data that was received.</param>
+        void OnIncomingNetMessage( EMsg msgType, byte[] data );
```
Comment (Netshroud) [inline SteamKit2/SteamKit2/Util/DebugNetworkListener.cs]: I don't like the 'Net' shorthand, but meh.

File: Samples/3.DebugLog/Program.cs (hunk 5 above)
Comment (xPaw) [inline Samples/3.DebugLog/Program.cs]: Perhaps commenting it out by default would be better?

File: SteamKit2/SteamKit2/Util/DebugNetworkListener.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+            LogDirectory = path;
+        }
+
+        /// <summary>
+        /// Called when a packet is received from the Steam server.
+        /// </summary>
+        /// <param name="msgType">Network message type of this packet message.</param>
+        /// <param name="data">Raw packet data that was received.</param>
+        public void OnIncomingNetMessage( EMsg msgType, byte[] data )
```
Comment (Netshroud) [inline SteamKit2/SteamKit2/Util/DebugNetworkListener.cs]: Seems like a good candidate for an expression-bodied method, but again meh.

File: SteamKit2/Tests/NetworkListenerFacts.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
+﻿using Xunit;
+using SteamKit2;
+using SteamKit2.Internal;
+
+namespace Tests
+{
+    class TestNetworkListener : IDebugNetworkListener
+    {
+        public EMsg NextExpectedSentMessage;
```
Comment (Netshroud) [inline SteamKit2/Tests/NetworkListenerFacts.cs]: This is backwards from how I've usually seen tests like this - store the last one in the stub, assert from the test itself. Here you have no control from the test itself about where this runs.

File: SteamKit2/Tests/NetworkListenerFacts.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        public void OnIncomingNetMessage(EMsg msgType, byte[] data)
+        {
+            //
+        }
+
+        public void OnOutgoingNetMessage(EMsg msgType, byte[] data)
+        {
+            var expected = NextExpectedSentMessage;
+
+            NextExpectedSentMessage = 0;
```
Comment (Netshroud) [inline SteamKit2/Tests/NetworkListenerFacts.cs]: This doesn't require a cast?

File: SteamKit2/Tests/NetworkListenerFacts.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+            Assert.Equal(expected, msgType);
```
Comment (Netshroud) [inline SteamKit2/Tests/NetworkListenerFacts.cs]: Assert data too - maybe just a simple test, such as not null and length > 0

File: SteamKit2/Tests/NetworkListenerFacts.cs (hunk 9 above)
Comment (xPaw) [inline SteamKit2/Tests/NetworkListenerFacts.cs]: Doesn't appear to.

File: Samples/3.DebugLog/Program.cs (hunk 5 above)
Comment (Netshroud) [inline Samples/3.DebugLog/Program.cs]: I'd probably make it a separate sample with warnings everywhere, or leave it unsampled. It seems fairly simple enough to add into any pre-existing sample.

File: SteamKit2/SteamKit2/Steam/CMClient.cs [hunk 11]
```diff
@@ -194,6 +199,7 @@ public void Send( IClientMsg msg )
 
             DebugLog.WriteLine( "CMClient", "Sent -> EMsg: {0} (Proto: {1})", msg.MsgType, msg.IsProto );
 
+            DebugNetworkListener?.OnOutgoingNetMessage( msg.MsgType, msg.Serialize() );
```
Comment (Netshroud) [inline SteamKit2/SteamKit2/Steam/CMClient.cs]: What happens if the debug listener throws an exception? Same for incoming below.

Comment (xPaw) [inline SteamKit2/SteamKit2/Steam/CMClient.cs]: DebugLog can throw too, can't it?

File: SteamKit2/SteamKit2/Util/DebugNetworkListener.cs [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        /// Will create a folder in path "%assembly%/nethook/%currenttime%/"
+        /// </summary>
+        public NetHookNetworkListener()
+        {
+            var uri = new Uri( System.Reflection.Assembly.GetExecutingAssembly().CodeBase );
+
+            LogDirectory = Path.Combine(
+                Path.GetDirectoryName( uri.LocalPath ),
+                "nethook",
+                DateUtils.DateTimeToUnixTime( DateTime.Now ).ToString()
```
Comment (Netshroud) [inline SteamKit2/SteamKit2/Util/DebugNetworkListener.cs]: Seems a bit racy - two concurrently created `SteamClient`s with separate listeners could both try claim the same directory.

File: SteamKit2/SteamKit2/Steam/CMClient.cs (hunk 11 above)
Comment (Netshroud) [inline SteamKit2/SteamKit2/Steam/CMClient.cs]: Yep, but that doesn't make it right.

**Comment by Netshroud** [comment]
:+1:

**Pull request merged**